File: playwright-csharp/tests/Configuration/TestRunConfiguration.cs
using TestTracks.Playwright.CSharp.Configuration;
using TestTracks.Playwright.CSharp.Specs.Reporting;

namespace TestTracks.Playwright.CSharp.Specs.Configuration;

public sealed record TestRunContext(
    string EnvironmentName,
    TestSettings Settings,
    EnvironmentDataStore Data);

public static class TestRunConfiguration
{
    private const string DefaultEnvironment = "dev";
    private const string DefaultBrowser = "chromium";
    private const bool DefaultHeadless = false;

    private static readonly Lazy<TestRunContext> CurrentRun = new(Create);

    public static TestRunContext Current => CurrentRun.Value;

    private static TestRunContext Create()
    {
        var environmentName = GetRunValue("ENV", DefaultEnvironment);
        var browser = ParseBrowser(GetRunValue("BROWSER", DefaultBrowser));
        var headless = ParseBool(GetRunValue("HEADLESS", DefaultHeadless.ToString()));
        var data = new EnvironmentDataStore(environmentName);
        var targets = data.Load<EnvironmentTargets>("all-targets");
        var settings = targets.Settings;

        return new TestRunContext(
            environmentName,
            new TestSettings(
                browser,
                headless,
                settings.DefaultTimeoutMs,
                settings.ExpectTimeoutMs,
                settings.TraceOnFailure,
                RunArtifacts.Directory),
            data);
    }

    private static string GetRunValue(string name, string defaultValue)
    {
        var value = Environment.GetEnvironmentVariable(name);
        return string.IsNullOrWhiteSpace(value) ? defaultValue : value.Trim();
    }

    private static BrowserName ParseBrowser(string value)
    {
        return value.Trim().ToLowerInvariant() switch
        {
            "chromium" => BrowserName.Chromium,
            "firefox" => BrowserName.Firefox,
            "edge" => BrowserName.Edge,
            _ => throw new InvalidOperationException(
                $"Unsupported BROWSER value '{value}'. Use chromium, firefox or edge.")
        };
    }

    private static bool ParseBool(string value)
    {
        return bool.TryParse(value, out var parsed)
            ? parsed
            : throw new InvalidOperationException("HEADLESS must be true or false.");
    }
}

File: playwright-java/src/test/java/testtracks/configuration/TestRunConfiguration.java
package testtracks.configuration;

public final class TestRunConfiguration {
  private static final String DEFAULT_ENVIRONMENT = "dev";
  private static final String DEFAULT_BROWSER = "chromium";
  private static final boolean DEFAULT_HEADLESS = false;
  private static final TestRunContext CURRENT = create();

  private TestRunConfiguration() {}

  public static TestRunContext current() {
    return CURRENT;
  }

  private static TestRunContext create() {
    var environmentName = runValue("ENV", DEFAULT_ENVIRONMENT);
    var browser = parseBrowser(runValue("BROWSER", DEFAULT_BROWSER));
    var headless = parseBoolean(runValue("HEADLESS", Boolean.toString(DEFAULT_HEADLESS)));
    var data = new EnvironmentDataStore(environmentName);
    var targets = data.load("all-targets", EnvironmentTargets.class);
    var settings = targets.settings();

    return new TestRunContext(
        environmentName,
        new TestSettings(
            browser,
            headless,
            settings.defaultTimeoutMs(),
            settings.expectTimeoutMs(),
            settings.traceOnFailure(),
            RunArtifacts.directory()),
        data);
  }

  private static String runValue(String name, String defaultValue) {
    var propertyValue = System.getProperty(name);
    if (propertyValue != null && !propertyValue.isBlank()) {
      return propertyValue.trim();
    }

    var environmentValue = System.getenv(name);
    return environmentValue == null || environmentValue.isBlank()
        ? defaultValue
        : environmentValue.trim();
  }

  private static BrowserName parseBrowser(String value) {
    return switch (value.trim().toLowerCase()) {
      case "chromium" -> BrowserName.CHROMIUM;
      case "firefox" -> BrowserName.FIREFOX;
      case "edge" -> BrowserName.EDGE;
      default ->
          throw new IllegalStateException(
              "Unsupported BROWSER value '" + value + "'. Use chromium, firefox or edge.");
    };
  }

  private static boolean parseBoolean(String value) {
    if ("true".equalsIgnoreCase(value) || "false".equalsIgnoreCase(value)) {
      return Boolean.parseBoolean(value);
    }
    throw new IllegalStateException("HEADLESS must be true or false.");
  }
}
